src/OpenTelemetry.Extensions.Serilog/OpenTelemetrySerilogSink.cs

```diff
@@ -0,0 +1,135 @@
+
+            this.openTelemetryLoggerProvider = openTelemetryLoggerProvider!;
+            this.disposeProvider = disposeProvider;
+
+            var logEmitter = this.openTelemetryLoggerProvider.CreateEmitter();
+            Debug.Assert(logEmitter != null, "logEmitter was null");
+
+            this.logEmitter = logEmitter!;
+
+            // TODO: This project can only access IncludeFormattedMessage
+            // because it can see SDK internals. At some point this is likely
+            // not to be the case. Need to figure out where to put
+            // IncludeFormattedMessage so that extensions can see it. Ideas:
+            // Make it public on OpenTelemetryLoggerProvider or expose it on
+            // LogEmitter instance.
+            this.includeFormattedMessage = this.openTelemetryLoggerProvider.IncludeFormattedMessage;
+        }
+
+        public void Emit(LogEvent logEvent)
+        {
+            Debug.Assert(logEvent != null, "LogEvent was null.");
+
+            LogRecordData data = new(Activity.Current)
+            {
+                Timestamp = logEvent!.Timestamp.UtcDateTime,
+                LogLevel = (LogLevel)(int)logEvent.Level,
+                Message = this.includeFormattedMessage ? logEvent.RenderMessage() : logEvent.MessageTemplate.Text,
+                Exception = logEvent.Exception,
+            };
+
+            LogRecordAttributeList attributes = default;
+            foreach (KeyValuePair<string, LogEventPropertyValue> property in logEvent.Properties)
+            {
+                // TODO: Serilog supports complex type logging. This is not yet
+                // supported in OpenTelemetry.
+                if (property.Key == Constants.SourceContextPropertyName
+                    && property.Value is ScalarValue sourceContextValue)
+                {
+                    data.CategoryName = sourceContextValue.Value as string;
+                }
+                else if (property.Value is ScalarValue scalarValue)
+                {
+                    attributes.Add(property.Key, scalarValue.Value);
+                }
+                else if (property.Value is SequenceValue sequenceValue)
+                {
+                    IReadOnlyList<LogEventPropertyValue> elements = sequenceValue.Elements;
+                    if (elements.Count > 0)
+                    {
+                        // Note: The goal here is to build a typed array (eg
+                        // int[]) if all the element types match otherwise
+                        // fallback to object[]
+
+                        Type? elementType = null;
+                        Array? values = null;
+
+                        for (int i = 0; i < elements.Count; i++)
+                        {
+                            if (elements[i] is ScalarValue value)
+                            {
+                                Type currentElementType = value.Value?.GetType() ?? typeof(object);
```

Comment:
If this were of type `MyDisposableType` then later when this log record is exported it'd throw an object disposed exception when attempting to ToString the object. The exporters catch and swallow this kind of thing today.

We don't do anything special earlier in the pipeline for attributes on traces or metrics, but is there any reason we'd want to consider coercing things into one of the primitive types earlier for log data?

Reply:
Not totally following you. But I don't think a complex type would end up as a `ScalarValue`.

I have this test at the moment:

https://github.com/open-telemetry/opentelemetry-dotnet/blob/1643b316971a3b45302046ad4dc680efa69763ee/test/OpenTelemetry.Extensions.Serilog.Tests/OpenTelemetrySerilogSinkTests.cs#L155

That is verifying a ["destructed"](https://github.com/serilog/serilog/wiki/Structured-Data#preserving-object-structure) complex type is currently dropped on the floor, because we don't yet have support for that. What happens in that case is instead of a `ScalarValue` you get a `StructureValue` which itself contains all the properties of the complex type represented as `ScalarValue`s or other `StructureValue`s, etc. I don't think disposable/mutable things are a concern because Serilog will capture all the data from the complex type and then represent it using these "value" types it has in here: https://github.com/serilog/serilog/tree/dev/src/Serilog/Events

Right now we have basic support. You can use Serilog to do structured log messaging with primitives + arrays of primitives and it will work into OTel. My goal is to get support into OTel _somehow_ for complex types and then come back to this and properly support fully Serilog's destructuring mechanism 😄

Reply:
Ah thanks I missed that aspect of the test. Definitely seems like a bridge to cross later once complex types are supported. If serilog does all the work then my question is moot. I wasn't sure if a complex type could end up being added to the attributes of a log record.